File: cal_proot.py
"""
An efficient solution of finding Primitive Root is based on the below facts. 
If the multiplicative order of a number r modulo n is equal to Euler Totient Function ?(n) 
( note that the Euler Totient Function for a prime n is n-1), then it is a primitive root. 
e.g.,
P = 7
Euler_Totient_Function(7) = 6
we have 3^6 = 1 (mod 7)
"""
import math

def is_prime(num):
    # print("* checking prime")
    if num <= 1:
        return False
    
    for i in range(2, int(math.sqrt(num))+1):
        if (num % i) == 0:
        #    print(num, "is not a prime")
        #    print(i,"乘于",num//i,"是",num)
            return False
    # print(num, "is a prime")
    return True


def qpow(a, n, p):
    """a^n mod p"""
    ans = 1
    a = (a % p + p) % p
    while n:
        if n & 1:
            ans = (a * ans) % p
        a = (a * a) % p
        n >>= 1
    return ans

# 用辗转相除求最大公因子
def gcd(a, b):
    r = a % b
    while(r != 0):
        a = b
        b = r
        r = a % b
    return b

# 欧拉函数-暴力循环版
def euler(a):
    print("* cal euler")
    if is_prime(a):
        return a-1
    count = 0
    for i in range(1, a):
        if gcd(a, i) == 1:
            count += 1
    return count

def get_min_order(g, p, n):
    for i in range(1, n+1):
        if i % 1000000 == 0:
            print(f"** cal order {i} ...")
        if qpow(g, i, p) == 1:
            return i
    return 0

def primitive_root(p, get_min=True):
    """order(g, p) == euler(p)"""
    n = euler(p)
    print("Eular:", n)
    primitive_roots = []
    for g in range(2, p):
        print("g:", g)
        if gcd(g, p) != 1:
            continue
        if get_min_order(g, p, n) == n:
            print("find proot:", g)
            primitive_roots.append(g)
            if get_min:
                break
        exit()
    return primitive_roots

def get_prime_factors(n):
    factors = []
    x = 2
    next_prime = False
    while(x<=n):
        # print(n, x)
        if n % x == 0:
            if x not in factors:
                factors.append(x)
            n = n // x
        else:
            next_prime = True
        # get next prime
        while(next_prime or not is_prime(x)):
            x += 1
            next_prime = False

    return factors

def fast_min_primitive_root(p):
    """only work for prime number"""
    assert(is_prime(p))

    s = p - 1
    factors = get_prime_factors(s)
    ts = [s//v for v in factors]
    for g in range(2, p):
        # actually you can skip g=g_prev^2
        cnt = 0
        for t in ts:
            if qpow(g, t, p) == 1:
                break
            else:
                cnt += 1
        if cnt == len(ts):
            return g
    return 0

if __name__ == "__main__":
    P = 998244353
    # P = 68719403009
    # P = 68719230977
    # P = 137438822401 
    # root = primitive_root(P, get_min=True)
    root = fast_min_primitive_root(P)
    print(root)

File: cal_proot.cpp
/* 
C++ program to find the primitive root of p
If the multiplicative order of a number r modulo n is equal to Euler Totient Function ?(n) 
( note that the Euler Totient Function for a prime n is n-1), then it is a primitive root. 
e.g.,
P = 7
Euler_Totient_Function(7) = 6
we have 3^6 = 1 (mod 7)
*/
#include <iostream>
#include <math.h>
#include <vector>
using namespace std;

typedef __int128_t int128;
typedef unsigned long long uint64;

inline void printInt128(int128 x)
{
    if (x < 0)
    {
        putchar('-');
        x = -x;
    }
    if (x > 9)
        printInt128(x / 10);
    putchar(x % 10 + '0');
}

// Returns true if n is prime
template <class T>
T isPrime(T n)
{
    cout << "* Checking prime" << endl;
    // Corner cases
    if (n <= 1)
        return false;
    if (n <= 3)
        return true;

    // This is checked so that we can skip
    // middle five numbers in below loop
    if (n % 2 == 0 || n % 3 == 0)
        return false;

    for (T i = 5; i * i <= n; i = i + 6)
    {
        if (n % i == 0 || n % (i + 2) == 0)
        {
            // cout << i << " or " << (i+2) << endl;
            return false;
        }
    }

    return true;
}

/* Iterative Function to calculate (x^y)%p in O(logy) */
template <class T>
T qpow(T x, T y, T p)
{
    T res = 1; // Initialize result
    x = x % p;

    while (y != 0)
    {
        // If y is odd, multiply x with result
        if (y & 1)
            res = (res * x) % p;

        y = y >> 1;
        x = (x * x) % p;
    }
    return res;
}

template <class T>
T gcd(T a, T b)
{
    T r = a % b;
    while (r != 0)
    {
        a = b;
        b = r;
        r = a % b;
    }
    return b;
}

template <class T>
T eular(T x)
{
    if (isPrime(x))
    {
        cout << "a prime" << endl;
        return x - 1;
    }
    cout << "not a prime" << endl;

    T cnt = 0;
    for (T i = 1; i < x; i++)
    {
        if (gcd(i, x) == 1)
        {
            cnt += 1;
        }
    }
    return cnt;
}

template <class T>
T getMinOrder(T g, T n, T p)
{
    for (T i = 1; i < n + 1; i++)
    {
        if (i % 1000000000 == 0)
        {
            printInt128(i);
            cout << " ..." << endl;
            // cout << "** cal order " << i << "..." << endl;
        }
        if (qpow(g, i, p) == 1)
            return i;
    }
    return 0;
}

template <class T>
vector<T> findPrimitiveRoot(T x, bool get_min = true)
{
    T n = eular(x);
    // cout << "Eular: "<< n << endl;
    vector<T> roots;
    for (T g = 2; g < x; g++)
    {
        cout << "g: ";
        printInt128(g);
        cout << endl;
        if (gcd(g, x) != 1)
            continue;
        if (getMinOrder(g, n, x) == n)
        {
            roots.push_back(g);
            if (get_min)
                return roots;
        }
    }

    return roots;
}

template <class T>
void printResult(T P, vector<T> roots)
{
    cout << "Get primitive root of " << P << ":" << endl;
    for (T x : roots)
        cout << x << " ";
    cout << endl;
}

void printResultInt128(int128 P, vector<int128> roots)
{
    cout << "Get primitive root of ";
    printInt128(P);
    cout << ":" << endl;
    for (int128 x : roots)
    {
        printInt128(x);
        cout << endl;
    }
}

int main()
{
    // uint64 P = 31;
    // uint64 P = 998244353; // 30bits
    // vector<uint64> roots = findPrimitiveRoot(P, true);
    // printResult(P, roots);

    int128 P = 68719403009; // 36bits
    // int128 P = 68719230977; // 36bits
    // int128 P = 137438822401; // 37bits
    vector<int128> roots = findPrimitiveRoot(P, true);
    printResultInt128(P, roots);

    return 0;
}